### PR TITLE
chunk_by_title decorator

### PR DESCRIPTION
**Pull request by Coniferish**
### Summary
Partial solution to #1185.
Related to #1222.
Creates decorator from `chunk_by_title` cleaning brick.
Breaks a document into sections based on the presence of Title elements. Also starts a new section under the following conditions:

- If metadata changes, indicating a change in section or page or a switch to processing attachments. If `multipage_sections=True`, sections can span pages. `multipage_sections` defaults to True.
- If the length of the section exceeds `new_after_n_chars` characters. The default is 1500. The **chunking function does not split individual elements**, so it's possible for a section to exceed that threshold if an individual element if over `new_after_n_chars characters`, which could occur with a long NarrativeText element.

Combines sections under these conditions
- Sections under `combine_under_n_chars` characters are combined. The default is 500.

### Testing
```
from unstructured.partition.html import partition_html

url = "https://understandingwar.org/backgrounder/russian-offensive-campaign-assessment-august-27-2023-0"
chunks = partition_html(url=url, chunking_strategy="by_title")

for chunk in chunks:
    print(chunk)
    print("\n\n" + "-"*80)
    input()
```

### CODE REVIEW

**Comment by Coniferish** [comment]
@MthwRobinson and/or @cragwolfe
I still need to add tests, but wanted to check in to make sure this is looking good so far.

**Comment by cragwolfe** [comment]
> @MthwRobinson and/or @cragwolfe
I still need to add tests, but wanted to check in to make sure this is looking good so far.

the pattern generally looks good to me. i'll observe, though, that we have a lot of metadata fields floating around now (somehave more doc types that others, e.g. "emphasized_text" and "href's"), and merging will be non-trivial.
what should happen when `page_number` is multiple pages?

**Comment by Coniferish** [comment]
> emphasized_text

@cragwolfe  Yah, I've noticed that as well and that's actually part of the reason why I put the docstring info for this in the decorator. 

I'm not clear on your question about `page_number`. Are you asking what if multiple sections have the same `page_number` or when a section spans multiple pages? I admittedly didn't think about that since I was just relying on Matt's function. I can investigate, though

**Comment by cragwolfe** [comment]
@Coniferish , please add a test for `partition()` from `unstructured.partition.auto` as well.

File: unstructured/chunking/title.py [hunk 1]
```diff
@@ -33,6 +35,17 @@ def chunk_by_title(
...
+    if (
+        combine_under_n_chars is not None
+        and new_after_n_chars is not None
+        and (
+            combine_under_n_chars > new_after_n_chars
+            or combine_under_n_chars < 0
+            or new_after_n_chars < 0
+        )
+    ):
+        raise ValueError("Invalid values for combine_under_n_chars and/or new_after_n_chars.")
```
Comment (cragwolfe) [inline unstructured/chunking/title.py:47]: please add a unittest for this

**Comment by cragwolfe** [comment]
Please add unittests for all the partition_ functions for the sake of testing completeness (even though logically, "should be fine").

**Comment by cragwolfe** [comment]
> > the pattern generally looks good to me. i'll observe, though, that we have a lot of metadata fields floating around now (somehave more doc types that others, e.g. "emphasized_text" and "href's"), and merging will be non-trivial. what should happen when `page_number` is multiple pages?
> 
> @cragwolfe Yah, I've noticed that as well and that's actually part of the reason why I put the docstring info for this in the decorator. What do you mean by "merging will be non-trivial" though?
> 
> I'm not clear on your question about `page_number`. Are you asking what if multiple sections have the same `page_number` or when a section spans multiple pages? I admittedly didn't think about that since I was just relying on Matt's function. I can investigate, though

we can kick this off to a separate issue -- comprehensive review of how to handle metadata merges when chunking.

**Review by cragwolfe — APPROVED** [review]

**Pull request merged**